File: quick.library/src/main/java/org/chris/quick/widgets/SemicircleView.java
package org.chris.quick.widgets;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.RadialGradient;
import android.graphics.Shader;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.View;

import org.chris.quick.R;
import org.chris.quick.b.BaseApplication;

/**
 * Created by work on 2017/4/20.
 * 半圆-红包上半截
 *
 * @author chris zou
 * @mail chrisSpringSmell@gmail.com
 */

public class SemicircleView extends View {
    public static final double BASE_WIDTH = 720;
    float scale;
    Paint mPaint;
    float radius;

    public SemicircleView(Context context) {
        this(context, null);
    }

    public SemicircleView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        TypedArray ta = context.obtainStyledAttributes(attrs, R.styleable.SemicircleView);
        radius = ta.getDimension(R.styleable.SemicircleView_SemicircleViewRadius, 0);
        ta.recycle();
        init();
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        int width = MeasureSpec.getSize(widthMeasureSpec);
        int height = (int) (MeasureSpec.getSize(heightMeasureSpec) * scale);
        setMeasuredDimension(width, height);
    }

    private void init() {
        scale = (float) BaseApplication.width / (float) BASE_WIDTH;//换算为指定比例
        radius *= scale;
        mPaint = new Paint();
        mPaint.setAntiAlias(true);
        mPaint.setColor(Color.BLACK);
    }

    @Override
    public void draw(Canvas canvas) {
        super.draw(canvas);
        drawCirle(canvas);
    }

    private void drawCirle(Canvas canvas) {
        int[] colors = new int[]{Color.parseColor("#f55a67"), Color.parseColor("#f44b55")};
        RadialGradient linearGradient = new RadialGradient(getWidth() / 2, 0, getHeight() / 2, colors, null, Shader.TileMode.MIRROR);
        mPaint.setShader(linearGradient);
        mPaint.setShadowLayer(10 * scale, 0, 2 * scale, Color.BLACK);
        float startX = radius, startY = 0;
        float endX = getWidth(), endY = radius;

        Path path = new Path();
        path.moveTo(startX, startY);
        if (radius > 0) {
            path.quadTo(0, 0, 0, radius);
        }
        //
        path.quadTo((endX - startX) / 2, getHeight(), endX, endY);
        path.moveTo(endX, endY);
        if (radius > 0) {
            path.quadTo(endX, endY - radius, endX - radius, 0);
        }
        path.lineTo(startX, startY);
        canvas.drawPath(path, mPaint);
    }
}

File: quick.library/src/main/java/org/chris/quick/widgets/SemicircleView.kt
package org.chris.quick.widgets

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RadialGradient
import android.graphics.Shader
import android.util.AttributeSet
import android.view.View
import com.zhy.autolayout.utils.AutoUtils
import org.chris.quick.R


/**
 * Created by work on 2017/4/20.
 *
 * @author chris zou
 * @mail chrisSpringSmell@gmail.com
 */

class SemicircleView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null) : View(context, attrs) {
    internal var width: Int = 0
    internal var height: Int = 0
    lateinit var mPaint: Paint
    var radius: Float = 0.toFloat()
    var shadowDistance = 0f
    var shadowRadious = 0f
    var startColor: Int = 0
    var endColor: Int = 0
    var paddingHeight: Float = 0.toFloat()
    var isShadow = true

    init {
        init()
        val ta = context.obtainStyledAttributes(attrs, R.styleable.SemicircleView)
        radius = ta.getDimension(R.styleable.SemicircleView_svRadius, 0f)
        startColor = ta.getColor(R.styleable.SemicircleView_svStartColor, Color.BLACK)
        endColor = ta.getColor(R.styleable.SemicircleView_svEndColor, Color.BLACK)
        paddingHeight = ta.getDimension(R.styleable.SemicircleView_svPaddingHeight, 0.0f)
        isShadow = ta.getBoolean(R.styleable.SemicircleView_svIsShadow, true)
        shadowDistance = ta.getDimension(R.styleable.SemicircleView_svShadowDistance, AutoUtils.getPercentHeightSize(10).toFloat())
        shadowRadious = ta.getDimension(R.styleable.SemicircleView_svShadowRadius, AutoUtils.getPercentWidthSize(2).toFloat())
        ta.recycle()
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        width = w
        height = h
        if (paddingHeight > h) {
            paddingHeight = h.toFloat()
        }
    }

    private fun init() {
        mPaint = Paint()
        mPaint.isAntiAlias = true
        mPaint.color = Color.BLACK
    }

    fun setColors(vararg colors: Int) {
        startColor = colors[0]
        endColor = if (colors.size >= 2)
            colors[1]
        else
            startColor
        postInvalidate()
    }

    override fun draw(canvas: Canvas?) {
        super.draw(canvas)
        configPaint()
        drawLine(canvas)
    }

    private fun configPaint() {
        if (endColor == Color.BLACK) {
            mPaint.color = startColor
        } else {
            val colors = intArrayOf(startColor, endColor)
            val linearGradient = RadialGradient((getWidth() / 2).toFloat(), 0f, (height / 2).toFloat(), colors, null, Shader.TileMode.MIRROR)
            mPaint.shader = linearGradient
        }
        if (isShadow) {
            mPaint.setShadowLayer(shadowRadious, 0f, shadowDistance, Color.BLACK)
        }
    }

    private fun drawLine(canvas: Canvas?) {
        val path = Path()
        path.moveTo(radius, 0f)
        if (radius > 0) {
            path.quadTo(0f, 0f, 0f, radius)
        }
        path.lineTo(0f, height.toFloat() - paddingHeight)
//        path.quadTo(width / 2.0f, height.toFloat() , width.toFloat(), height.toFloat() - paddingHeight)
        path.quadTo(width / 2.0f, height.toFloat() + paddingHeight, width.toFloat(), height.toFloat() - paddingHeight)
        path.lineTo(width.toFloat(), radius)
        if (radius > 0)
            path.quadTo(width.toFloat(), 0f, width.toFloat() - radius, 0f)
        path.lineTo(0f, 0f)
        canvas?.drawPath(path, mPaint)
    }
}
